core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -3911,14 +3918,21 @@ void checkAnd(SqlCall call) {
       }
     }
 
-    @Override public @Nullable SqlNode visit(final org.apache.calcite.sql.SqlCall call) {
+    @Override public @Nullable SqlNode visit(final SqlCall call) {
       SqlKind kind = call.getKind();
-      if (kind != SqlKind.AND && kind != SqlKind.EQUALS) {
+      if (kind != SqlKind.AND && kind != SqlKind.EQUALS && kind != SqlKind.CAST) {
         illegal = true;
```

Comment:
I check the `checkAnd(call)` and this method, Maybe we can exit directly when we find the illegal value is true.

Reply:
Some minor advice:
1) SqlValidatorImpl#validateJoin(Line#3784). There is an extra `{}` behind the ASOF.
2) ConjunctionOfEqualities#checkAnd method's java comment. Before `Children`  There is an extra space.

Reply:
the extra {} are actually useful, they create a local scope, and the variables in the local scope do not interfere with variables from the rest of the function. It also prevents accidental updates to other variables that may be in scope. If also helps with the debugger too: such variables go out of scope when the inner scope {} terminates, so in large functions with lots of variables it makes the debugging experience better. I use this trick a lot when I have to write very large functions.